### PR TITLE
Mente registration flow

### PR DESCRIPTION
**Pull request by DaniWitcher**
1. Root component
2. Registration component
3. Registration guard
4. Mentee reactive form
5. Mentee form validation

### CODE REVIEW

File: projects/core/src/lib/authorization/components/auth/auth.component.html [hunk 1]
```diff
@@ -0,0 +1 @@
+<router-outlet></router-outlet>
```
Comment (Tsurhanau) [inline projects/core/src/lib/authorization/components/auth/auth.component.html:1]: move to RootComponent

File: projects/common-components/src/lib/mentee-form/mentee-form.component.html [hunk 2]
```diff
@@ -0,0 +1,63 @@
+<div class="mentee-form">
+	<form [formGroup]="menteeRegistrationForm">
+		<div class="mentee-form__components">
+			<div class="mentee-form__component">
+				<label class="mentee-form__label">Enter the name:</label>
+				<input class="mentee-form__input" formControlName="name" required />
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/common-components/src/lib/mentee-form/mentee-form.component.html:6]: Done

File: projects/core/src/lib/authorization/components/register/register.component.html [hunk 3]
```diff
@@ -0,0 +1 @@
+<p>register works!</p>
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/core/src/lib/authorization/components/register/register.component.html:1]: Done

File: src/app/app-routing.module.ts [hunk 4]
```diff
@@ -6,18 +6,27 @@ import { RoleSelectComponent } from "../../projects/core/src/lib/authorization/c
...
+	{ path: "role-select", component: RoleSelectComponent },
 	{ path: "talks", component: TalksListComponent },
-	{ path: "mentee-page", component: MenteePageComponent, canActivate: [AuthGuard] }
+	{ path: "mentee-page", component: MenteePageComponent },
+	{ path: "registration", component: RegisterComponent, canActivate: [RegisterGuard] }
+];
+
+const appRoutes: Routes = [
+	{ path: "", redirectTo: "/home-page", pathMatch: "full" },
+	{ path: "", component: AuthComponent, canActivate: [AuthGuard] },
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply src/app/app-routing.module.ts:25]: Done

File: src/app/app-routing.module.ts [hunk 5]
```diff
@@ -6,18 +6,27 @@ import { RoleSelectComponent } from "../../projects/core/src/lib/authorization/c
...
+	{ path: "", component: AuthComponent, children: routes, canActivate: [AuthGuard] }
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply src/app/app-routing.module.ts:26]: Done

File: projects/core/src/lib/authorization/components/auth/auth.component.html (hunk 1 above)
Comment (DaniWitcher) [reply projects/core/src/lib/authorization/components/auth/auth.component.html:1]: Done

File: projects/common-components/src/lib/mentee-form/mentee-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+export class MenteeFormComponent {
+	@Output() cancelEvent: EventEmitter<void> = new EventEmitter<void>();
+	@Output() submitEvent: EventEmitter<void> = new EventEmitter<void>();
+	@Input() lessonsList!: Lesson[];
+	public menteeRegistrationForm: FormGroup = this.formBuilder.group({
+		name: [null, Validators.required],
+		mail: [null, [Validators.required, Validators.email]],
+		rmMail: [null, [Validators.required, Validators.email]],
+		modules: this.formBuilder.array([false, false, false, false, false, false, false, false])
+	});
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/common-components/src/lib/mentee-form/mentee-form.component.ts]: Done

File: projects/common-components/src/lib/mentee-form/mentee-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	public checkAndSubmit(): void {
+		if (this.menteeRegistrationForm.valid && this.checkBoxValidator()) {
+			//toDo Save Data to DataBase
+		}
+		this.submitEvent.emit();
+	}
+
+	private checkBoxValidator(): boolean {
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/common-components/src/lib/mentee-form/mentee-form.component.ts]: Done

File: src/app/app-routing.module.ts [hunk 8]
```diff
@@ -5,24 +5,33 @@ import { AuthGuard } from "projects/core/src/lib/authorization/guards/auth/auth.
...
 			import("../../projects/core/src/lib/components/talks-page/talks-page.module").then(
 				(m) => m.TalksPageModule
 			)
-	},
-	{ path: "mentee-page", component: MenteePageComponent, canActivate: [AuthGuard] }
+	}
+];
+
+const appRoutes: Routes = [
+	{ path: "", component: RootComponent, canActivate: [AuthGuard] },
```
Comment (Tsurhanau) [inline src/app/app-routing.module.ts:29]: remove duplicate

File: src/app/app-routing.module.ts [hunk 9]
```diff
@@ -5,24 +5,33 @@ import { AuthGuard } from "projects/core/src/lib/authorization/guards/auth/auth.
...
+import { RegisterComponent } from "../../projects/core/src/lib/authorization/components/register/register.component";
+import { TalksListComponent } from "@jsmu/common-components";
 // route guards
 
 const routes: Routes = [
-	{ path: "", redirectTo: "/home-page", pathMatch: "full" },
 	{ path: "home-page", component: HomePageComponent },
-	{ path: "sign-in", component: SignInComponent },
-	{ path: "role-select", component: RoleSelectComponent, canActivate: [AuthGuard] },
+	{ path: "role-select", component: RoleSelectComponent },
```
Comment (Tsurhanau) [inline src/app/app-routing.module.ts:15]: checek and remove

File: projects/core/src/lib/services/users.data.service.ts [hunk 10]
```diff
@@ -34,7 +37,7 @@ export class UsersDataService {
 		);
 	}
 
-	public isUserLogin(): Observable<User | null> {
+	public isUserLogin(): Observable<User | null | undefined> {
```
Comment (Tsurhanau) [inline projects/core/src/lib/services/users.data.service.ts:40]: ?
!

File: projects/core/src/lib/models/talk.ts [hunk 11]
```diff
@@ -3,13 +3,13 @@ import { Status } from "./status";
 
 export interface Talk {
 	lessonId: number;
-	status: Status;
-	result: Result;
+	status: Status | null | "null";
```
Comment (Tsurhanau) [inline projects/core/src/lib/models/talk.ts:6]: null

File: projects/core/src/lib/models/mentee.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+export interface Mentee {
```
Comment (Tsurhanau) [inline projects/core/src/lib/models/mentee.ts:1]: !
?

File: projects/common-components/src/lib/mentee-form/mentee-form.component.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+	constructor(private formBuilder: FormBuilder, private talksService: TalksService) {
+		this.menteeRegistrationForm = this.formBuilder.group({
+			name: [null, Validators.required],
+			mail: [null, [Validators.required, Validators.email]],
+			rmMail: [null, [Validators.required, Validators.email]],
+			modules: this.formBuilder.array(
+				[false, false, false, false, false, false, false, false],
+				[Validators.required, checkBoxValidator()]
+			)
+		});
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/common-components/src/lib/mentee-form/mentee-form.component.ts:26]: Done

File: projects/common-components/src/lib/mentee-form/mentee-form.component.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	public cancelEmit(): void {
+		this.talksService.getTalksFromDb().subscribe((talk) => {
+			console.log(talk);
+		});
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/common-components/src/lib/mentee-form/mentee-form.component.ts:32]: Done

File: src/app/app.component.html [hunk 15]
```diff
@@ -1,5 +1,5 @@
 <jsmu-header></jsmu-header>
 <jsmu-info-panel *ngIf="isModalOpen$ | async"></jsmu-info-panel>
-<router-outlet></router-outlet>
+<jsmu-root-c></jsmu-root-c>
 <jsmu-dialog [dialogType]="dialogType$ | async" *ngIf="isOpen$ | async"></jsmu-dialog>
 <jsmu-footer></jsmu-footer>
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply src/app/app.component.html:5]: Done

File: src/app/app-routing.module.ts (hunk 8 above)
Comment (DaniWitcher) [reply src/app/app-routing.module.ts:29]: Done

File: src/app/app-routing.module.ts (hunk 9 above)
Comment (DaniWitcher) [reply src/app/app-routing.module.ts:15]: Done

File: projects/core/src/lib/models/talk.ts (hunk 11 above)
Comment (DaniWitcher) [reply projects/core/src/lib/models/talk.ts:6]: Done

File: projects/core/src/lib/models/mentee.ts (hunk 12 above)
Comment (DaniWitcher) [reply projects/core/src/lib/models/mentee.ts:1]: Done

File: projects/core/src/lib/services/users.data.service.ts (hunk 10 above)
Comment (DaniWitcher) [reply projects/core/src/lib/services/users.data.service.ts:40]: Done

File: projects/core/src/lib/services/users.data.service.ts [hunk 16]
```diff
@@ -17,7 +17,10 @@ export class UsersDataService {
 		return this.db.saveData<User>(ListType.USERS, newData);
 	}
 
-	public getUserById(uid: string): Observable<User | null> {
+	public getUserById(uid: string): Observable<User | null | undefined> {
+		if (!uid) {
+			return of(undefined);
+		}
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/core/src/lib/services/users.data.service.ts:23]: necessary as i mentioned in our call

File: projects/core/src/lib/authorization/guards/auth/auth.guard.ts [hunk 17]
```diff
@@ -1,21 +1,46 @@
...
+					.getUser()
+					.pipe(
+						mergeMap((user) => {
+							return this.userService.getUserById(user.uid!);
+						})
+					)
+					.subscribe((user) => {
+						if (user === null) {
+							this.router.navigate([Routes.REGISTRATION]);
+						} else if (user === undefined) {
```
[the remark this answers is not in the document]
Comment (DaniWitcher) [reply projects/core/src/lib/authorization/guards/auth/auth.guard.ts:37]: Done

File: projects/common-components/src/lib/talks/talk-item/talk-item.component.spec.ts [hunk 18]
```diff
@@ -1,24 +1,24 @@
...
+// 			schemas: [CUSTOM_ELEMENTS_SCHEMA]
+// 		}).compileComponents();
+//
+// 		fixture = TestBed.createComponent(TalkItemComponent);
+// 		component = fixture.componentInstance;
+// 		component.talk = talkMock;
+// 		fixture.detectChanges();
+// 	});
+//
+// 	it("should create", () => {
```
Comment (Tsurhanau) [inline projects/common-components/src/lib/talks/talk-item/talk-item.component.spec.ts:21]: fit xit

File: src/app/app-routing.module.ts [hunk 19]
```diff
@@ -2,27 +2,37 @@ import { NgModule } from "@angular/core";
...
 import { HomePageComponent } from "projects/core/src/lib/components/home-page/home-page.component";
+import { RootComponent } from "../../projects/core/src/lib/components/root/root.component";
 // route guards
 
 const routes: Routes = [
-	{ path: "", redirectTo: "/home-page", pathMatch: "full" },
 	{ path: "home-page", component: HomePageComponent },
-	{ path: "sign-in", component: SignInComponent },
-	{ path: "role-select", component: RoleSelectComponent, canActivate: [AuthGuard] },
+	//{ path: "talks", component: TalksListComponent },
```
Comment (Tsurhanau) [inline src/app/app-routing.module.ts:12]: remove

Comment (DaniWitcher) [reply src/app/app-routing.module.ts:12]: Done

File: projects/common-components/src/lib/talks/talk-item/talk-item.component.spec.ts (hunk 18 above)
Comment (DaniWitcher) [reply projects/common-components/src/lib/talks/talk-item/talk-item.component.spec.ts:21]: Doesnt change

**Review by Tsurhanau — APPROVED** [review]

**Pull request merged**